File: source/aoBlogs2/Models/Domain/ApplicationEnvironmentModel.vb

Imports System.Linq
Imports Contensive.Addons.Blog.Controllers
Imports Contensive.Addons.Blog.Models
Imports Contensive.BaseClasses

Namespace Models
    Public Class ApplicationEnvironmentModel
        '
        Public ReadOnly Property cp As CPBaseClass
        '
        '====================================================================================================
        ''' <summary>
        ''' The user at the keyboard is editing
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property userIsEditing As Boolean
            Get
                If local_userIsEditing Is Nothing Then
                    local_userIsEditing = cp.User.IsEditingAnything
                End If
                Return CBool(local_userIsEditing)
            End Get
        End Property
        Private Property local_userIsEditing As Boolean? = Nothing
        '
        '====================================================================================================
        ''' <summary>
        ''' blog set during construction
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property blog As BlogModel
            Get
                '
                ' -- blog set during object constructor
                Return local_blog
            End Get
        End Property
        Private local_blog As BlogModel = Nothing
        '
        '====================================================================================================
        ''' <summary>
        ''' Current Blog Entry. Returns null if no valid entry
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property blogEntry As BlogPostModel
            Get
                '
                ' -- return if set
                If (local_blogEntry IsNot Nothing) Then Return local_blogEntry
                '
                ' -- return null if blogentryid is not valid
                If (local_blogEntryId Is Nothing) OrElse (local_blogEntryId = 0) Then Return Nothing
                '
                ' -- blogEntryId is valid, return blog entry
                local_blogEntry = DbModel.create(Of BlogPostModel)(cp, cp.Utils.EncodeInteger(local_blogEntryId))
                If (local_blogEntry IsNot Nothing) Then Return local_blogEntry
                '
                ' -- if blogentryid is not 0, and blog is null, remove link alias that might have caused this
                LinkAliasController.deleteLinkAlias(cp, cp.Doc.PageId, CInt(local_blogEntryId))
                Return Nothing
            End Get
        End Property
        Private ReadOnly local_blogEntryId As Integer? = Nothing
        Private local_blogEntry As BlogPostModel = Nothing
        '
        '====================================================================================================
        ''' <summary>
        ''' The current user at the keyboard
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property user As PersonModel
            Get
                If (local_user Is Nothing) Then
                    local_user = PersonModel.create(cp, cp.User.Id)
                End If
                Return local_user
            End Get
        End Property
        Private local_user As PersonModel = Nothing
        '
        '====================================================================================================
        ''' <summary>
        ''' link to the blogs list page
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property blogBaseLink As String
            Get
                If (local_blogBaseLink IsNot Nothing) Then Return local_blogBaseLink
                local_blogBaseLink = cp.Content.GetPageLink(cp.Doc.PageId)
                Return local_blogBaseLink
            End Get
        End Property
        Private local_blogBaseLink As String = Nothing
        '
        '====================================================================================================
        ''' <summary>
        ''' RSS feed for this blog?
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property rssFeed As RSSFeedModel
            Get
                If (local_RSSFeed Is Nothing) Then
                    '
                    ' Get the Feed Args
                    local_RSSFeed = DbModel.create(Of RSSFeedModel)(cp, blog.RSSFeedID)
                    If (local_RSSFeed Is Nothing) Then
                        local_RSSFeed = DbModel.add(Of RSSFeedModel)(cp)
                        local_RSSFeed.name = blog.Caption
                        local_RSSFeed.description = "This is your First RssFeed"
                        local_RSSFeed.save(Of BlogModel)(cp)
                        blog.RSSFeedID = local_RSSFeed.id
                        blog.save(Of BlogModel)(cp)
                    End If
                End If
                Return local_RSSFeed
            End Get
        End Property
        Private local_RSSFeed As RSSFeedModel = Nothing
        '
        '====================================================================================================
        '
        Public ReadOnly Property blogPageBaseLink As String
            Get
                If local_blogPageBaseLink IsNot Nothing Then Return local_blogPageBaseLink
                local_blogPageBaseLink = cp.Content.GetLinkAliasByPageID(cp.Doc.PageId, "", "")
                Return local_blogPageBaseLink
            End Get
        End Property
        Private local_blogPageBaseLink As String = Nothing
        '
        '
        Public ReadOnly Property nextArticle As BlogPostModel
            Get
                If nextArticle_local IsNot Nothing Then Return nextArticle_local
                Dim postList As List(Of BlogPostModel) = DbModel.createList(Of BlogPostModel)(cp, "(blogID=" & blog.id & ")and(DateAdded>" & cp.Db.EncodeSQLDate(blogEntry.DateAdded) & ")", "DateAdded", 1, 1)
                If postList.Count = 0 Then Return Nothing
                nextArticle_local = postList.First()
                Return nextArticle_local
            End Get
        End Property
        Private nextArticle_local As BlogPostModel = Nothing
        '
        '
        Public ReadOnly Property nextArticleLink As String
            Get
                If nextArticle Is Nothing Then Return Nothing
                '
                Dim qs As String = LinkAliasController.getLinkAliasQueryString(cp, cp.Doc.PageId, nextArticle.id)
                nextArticleLink_local = cp.Content.GetPageLink(cp.Doc.PageId, qs)
                Return nextArticleLink_local
            End Get
        End Property
        Private nextArticleLink_local As String = Nothing
        '
        '
        Public ReadOnly Property nextArticleLinkCaption As String
            Get
                If nextArticle Is Nothing Then Return Nothing
                Return nextArticle.name
            End Get
        End Property

        '
        '====================================================================================================
        '
        Public Sub New(cp As CPBaseClass, blog As BlogModel, blogEntryId As Integer)
            '
            Me.cp = cp
            '
            local_blog = blog
            local_blogEntryId = blogEntryId
        End Sub
        '
        '
        ''' <summary>
        ''' if false, do now show tags. This blocks all the additional pages that hurt SEO
        ''' </summary>
        ''' <returns></returns>
        Public ReadOnly Property sitePropertyAllowTags As Boolean
            Get
                Return cp.Site.GetBoolean("blog allow tags", False)
            End Get
        End Property
    End Class
End Namespace

File: source/aoBlogs2/Models/Domain/ApplicationEnvironmentModel.cs

using System.Linq;
using Contensive.Blog.Controllers;
using Contensive.BaseClasses;
using Contensive.Models.Db;

namespace Contensive.Blog.Models {
    public class ApplicationEnvironmentModel {
        // 
        public CPBaseClass cp { get; private set; }
        // 
        // ====================================================================================================
        /// <summary>
        /// The user at the keyboard is editing
        /// </summary>
        /// <returns></returns>
        public bool userIsEditing {
            get {
                local_userIsEditing ??= cp.User.IsEditing();
                return (bool)local_userIsEditing;
            }
        }
        private bool? local_userIsEditing { get; set; } = default;
        // 
        // ====================================================================================================
        /// <summary>
        /// blog set during construction
        /// </summary>
        /// <returns></returns>
        public BlogModel blog {
            get {
                // 
                // -- blog set during object constructor
                return local_blog;
            }
        }
        private readonly BlogModel local_blog = null;
        // 
        // ====================================================================================================
        /// <summary>
        /// Current Blog Entry. Returns null if no valid entry
        /// </summary>
        /// <returns></returns>
        public BlogEntryModel blogPost {
            get {
                // 
                // -- return if set
                if (_blogPost is not null)
                    return _blogPost;
                // 
                // -- return null if blogentryid is not valid
                if (_blogPostId is null || (_blogPostId.HasValue ? _blogPostId.Value == 0 : (bool?)null).GetValueOrDefault())
                    return null;
                // 
                // -- blogEntryId is valid, return blog entry
                _blogPost = DbBaseModel.create<BlogEntryModel>(cp, cp.Utils.EncodeInteger(_blogPostId));
                if (_blogPost is not null) {
                    BlogEntryModel.verifyPost(cp, _blogPost);
                    return _blogPost;
                }
                // 
                // -- if blogentryid is not 0, and blog is null, remove link alias that might have caused this
                LinkAliasController.deleteLinkAlias(cp, (int)_blogPostId);
                return null;
            }
        }
        private readonly int? _blogPostId = default;
        private BlogEntryModel _blogPost = null;
        // 
        // ====================================================================================================
        /// <summary>
        /// The current user at the keyboard
        /// </summary>
        /// <returns></returns>
        public Contensive.Blog.Models.PersonModel user {
            get {
                local_user ??= PersonModel.create<PersonModel>(cp, cp.User.Id);
                return local_user;
            }
        }
        private PersonModel local_user = null;
        // 
        // ====================================================================================================
        /// <summary>
        /// link to the blogs list page
        /// </summary>
        /// <returns></returns>
        public string blogBaseLink {
            get {
                if (local_blogBaseLink is not null)
                    return local_blogBaseLink;
                local_blogBaseLink = cp.Content.GetPageLink(cp.Doc.PageId);
                return local_blogBaseLink;
            }
        }
        private string local_blogBaseLink = null;
        // 
        // ====================================================================================================
        /// <summary>
        /// RSS feed for this blog?
        /// </summary>
        /// <returns></returns>
        public RSSFeedModel rssFeed {
            get {
                if (local_RSSFeed is null) {
                    // 
                    // Get the Feed Args
                    local_RSSFeed = DbBaseModel.create<RSSFeedModel>(cp, blog.rssFeedId);
                    if (local_RSSFeed is null) {
                        local_RSSFeed = DbBaseModel.addDefault<RSSFeedModel>(cp);
                        local_RSSFeed.name = blog.caption;
                        local_RSSFeed.description = "This is your First RssFeed";
                        local_RSSFeed.save(cp);
                        blog.rssFeedId = local_RSSFeed.id;
                        blog.save(cp);
                    }
                }
                return local_RSSFeed;
            }
        }
        private RSSFeedModel local_RSSFeed = null;
        // 
        // ====================================================================================================
        // 
        public string blogPageBaseLink {
            get {
                if (local_blogPageBaseLink is not null)
                    return local_blogPageBaseLink;
                local_blogPageBaseLink = cp.Content.GetLinkAliasByPageID(cp.Doc.PageId, "", "");
                return local_blogPageBaseLink;
            }
        }
        private string local_blogPageBaseLink = null;
        // 
        // 
        public BlogEntryModel nextArticle {
            get {
                if (nextArticle_local is not null)
                    return nextArticle_local;
                var articleList = DbBaseModel.createList<BlogEntryModel>(cp, "(blogID=" + blog.id + ")and(dateAdded<" + cp.Db.EncodeSQLDate(cp.Utils.EncodeDate(blogPost.dateAdded)) + ")", "dateAdded desc", 1, 1);
                if (articleList.Count == 0) {
                    // 
                    // -- this may be the last article in the list, the next article should be the first to loop around
                    articleList = DbBaseModel.createList<BlogEntryModel>(cp, "(blogID=" + blog.id + ")", "dateAdded desc", 1, 1);
                    if (articleList.Count == 0)
                        return null;
                }
                nextArticle_local = articleList.First();
                return nextArticle_local;
            }
        }
        private BlogEntryModel nextArticle_local = null;
        // 
        // 
        public string nextArticleLink {
            get {
                if (nextArticle is null)
                    return null;
                // 
                string qs = LinkAliasController.getLinkAliasQueryString(cp, nextArticle.id);
                nextArticleLink_local = cp.Content.GetPageLink(cp.Doc.PageId, qs);
                return nextArticleLink_local;
            }
        }
        private string nextArticleLink_local = null;
        // 
        // 
        public string nextArticleLinkCaption {
            get {
                if (nextArticle is null)
                    return null;
                return nextArticle.name;
            }
        }

        // 
        // ====================================================================================================
        // 
        public ApplicationEnvironmentModel(CPBaseClass cp, BlogModel blog, int blogEntryId) {
            // 
            this.cp = cp;
            // 
            local_blog = blog;
            _blogPostId = blogEntryId;
        }
        // 
        // 
        /// <summary>
        /// if false, do now show tags. This blocks all the additional pages that hurt SEO
        /// </summary>
        /// <returns></returns>
        public bool sitePropertyAllowTags {
            get {
                return cp.Site.GetBoolean("blog allow tags", false);
            }
        }
    }
}